File: golang/y21/d1/d1.go
package d1

import (
	"fmt"

	"github.com/bobbykaz/advent-of-code/utilities"
)

func Part1() int {
	lines := utilities.ReadFileIntoLines("input/y21/d1.txt")
	numbers := utilities.StringsToInts(lines)

	prev := numbers[0]
	next := 0
	incs := 0
	for i := 1; i < len(numbers); i++ {
		next = numbers[i]
		if next > prev {
			incs++
		}
		prev = next
	}

	Part2(numbers)
	return incs
}

func Part2(nums []int) int {

	p1 := nums[0]
	p2 := nums[1]
	p3 := nums[2]
	psum := p1 + p2 + p3
	incs := 0
	p1 = p2
	p2 = p3
	for i := 3; i < len(nums); i++ {
		p3 = nums[i]
		sum := p1 + p2 + p3
		if sum > psum {
			incs++
		}
		p1 = p2
		p2 = p3
		psum = sum
	}
	fmt.Println("part 2 output:", incs)
	return incs
}

File: rust/src/years/y21/d1.rs
use crate::utilities::read_file_into_lines;
use crate::utilities::convert_lines_to_ints;

pub fn run() {
    let lines = read_file_into_lines("../input/y21/d1.txt");
    let nums = convert_lines_to_ints(&lines);
    pt1(&nums);
    pt2(&nums);
}

pub fn pt1(ints: &Vec<i32>) {
    let mut count = 0;

    for i in 1..ints.len() {
        if ints[i] > ints[i-1] {
            count+=1;
        }
    }

    println!("Number of increasing values: {count}")
}

pub fn pt2(ints: &Vec<i32>) {
    let mut sums: Vec<i32> = Vec::new();
    for i in 2..ints.len() {
        let sum = ints[i] + ints[i-1] + ints[i-2];
        sums.push(sum)
    }
    println!("Part 2:");
    pt1(&sums);
}